packages/grid/_modules_/grid/hooks/features/sorting/useSorting.ts

```diff
@@ -236,6 +236,14 @@ export const useSorting = (apiRef: ApiRef) => {
   const sortApi: SortApi = { getSortModel, setSortModel, onSortModelChange, applySorting };
   useApiMethod(apiRef, sortApi, 'SortApi');
 
+  React.useEffect(() => {
+    // When the rows prop change, we reset the sortedRows state.
+    setGridState((state) => ({
+      ...state,
+      sorting: { ...state.sorting, sortedRows: rowsProp.map((row) => row.id) },
```

Comment:
> At least we fixed the reported bug and we have a solid test in place, so why not defer a better solution to later. I suspect it will require a refactoring of the state. There is an issue with the control flow somewhere, it could be imperative calls that shouldn't be here, it could be missing dependencies, a desynchronization, no idea :).

I will fix it ;)

Reply:
Fixed